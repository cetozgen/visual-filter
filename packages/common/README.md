# visual-filter/common

Shared utilities for the @visual-filter packages
