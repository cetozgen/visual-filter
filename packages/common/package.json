{
  "name": "@visual-filter/common",
  "version": "1.0.3",
  "description": "Shared utilities for the @visual-filter packages",
  "keywords": [
    "visual-filter"
  ],
  "homepage": "https://github.com/obadakhalili/vue-visual-filter.git",
  "repository": {
    "type": "git",
    "url": "https://github.com/obadakhalili/vue-visual-filter.git"
  },
  "license": "MIT",
  "author": "Obada Khalili",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "clean": "rm -rf dist",
    "dev": "yarn build --watch",
    "format": "yarn run prettier --ignore-path ../../.gitignore --write",
    "preversion": "yarn build"
  },
  "devDependencies": {
    "prettier": "2.7.1",
    "typescript": "4.7.4"
  }
}
