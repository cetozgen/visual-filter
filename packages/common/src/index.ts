export enum FilterType {
  GROUP = "group",
  CONDITION = "condition",
}

export enum GroupType {
  AND = "and",
  NOT_AND = "not and",
  OR = "or",
  NOT_OR = "not or",
}

export enum DataType {
  NUMERIC = "numeric",
  NOMINAL = "nominal",
}

export function deepCopy(src: any): any {
  if (Object(src) !== src) {
    return src
  }
  if (src.constructor === Object) {
    return Object.entries(src).reduce(
      (objCopy, [key, value]) => ({ ...objCopy, [key]: deepCopy(value) }),
      {},
    )
  }
  return src.map(deepCopy)
}
