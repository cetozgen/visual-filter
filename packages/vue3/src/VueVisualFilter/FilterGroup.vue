<script>
export default {
  name: "FilterGroup",
  emits: ["addFilter", "deleteGroup"],
  props: {
    group: {
      type: Object,
      required: true,
      validator(value) {
        return value.constructor === Object
      },
    },
    filterTypes: {
      type: Array,
      required: true,
    },
    groupTypes: {
      type: Array,
      required: true,
    },
    removable: {
      type: Boolean,
      required: true,
    },
  },
  methods: {
    addFilter(newFilterType) {
      if (this.filterTypes.includes(newFilterType)) {
        this.$emit("addFilter", this.group.filters, newFilterType)
      }
    },
  },
}
</script>

<template>
  <div class="space-y-4">
    <div class="space-x-2">
      <slot name="groupTypes" v-bind="{ groupTypes, group }">
        <select v-model="group.groupType">
          <option v-for="type in groupTypes" :key="type" :value="type">
            {{ type }}
          </option>
        </select>
      </slot>
      <slot name="filterAddition" v-bind="{ filterTypes, addFilter }">
        <select @change="addFilter($event.target.value)">
          <option v-for="type in filterTypes" :key="type" :value="type">
            {{ type }}
          </option>
        </select>
      </slot>
      <slot
        v-if="removable"
        name="groupDeletion"
        :deleteGroup="() => $emit('deleteGroup', group)"
      >
        <button @click="$emit('deleteGroup', group)">x</button>
      </slot>
    </div>
    <div v-if="group.filters.length" class="ml-10 space-y-1">
      <slot name="groupChildren"></slot>
    </div>
  </div>
</template>
