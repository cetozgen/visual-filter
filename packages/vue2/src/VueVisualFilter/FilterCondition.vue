<script>
import { DataType } from "@visual-filter/common"

export default {
  name: "FilterCondition",
  emits: ["updateField", "deleteCondition"],
  props: {
    condition: {
      type: Object,
      required: true,
      validator(value) {
        return value.constructor === Object
      },
    },
    fieldNames: {
      type: Array,
      required: true,
    },
    numericMethodNames: {
      type: Array,
      required: true,
    },
    nominalMethodNames: {
      type: Array,
      required: true,
    },
  },
  computed: {
    isNumeric() {
      return this.condition.dataType === DataType.NUMERIC
    },
  },
  methods: {
    updateField(newFieldName) {
      if (this.fieldNames.includes(newFieldName)) {
        this.$emit("updateField", this.condition, newFieldName)
      }
    },
  },
}
</script>

<template>
  <div class="space-x-2">
    <slot name="fieldUpdation" v-bind="{ fieldNames, condition, updateField }">
      <select
        v-model="condition.fieldName"
        @change="updateField($event.target.value)"
      >
        <option v-for="field in fieldNames" :key="field" :value="field">
          {{ field }}
        </option>
      </select>
    </slot>
    <slot
      name="methodUpdation"
      v-bind="{
        numericMethodNames: isNumeric && numericMethodNames,
        nominalMethodNames: isNumeric || nominalMethodNames,
        condition,
      }"
    >
      <select v-model="condition.method">
        <option
          v-for="method in isNumeric ? numericMethodNames : nominalMethodNames"
          :key="method"
          :value="method"
        >
          {{ method }}
        </option>
      </select>
    </slot>
    <slot name="argumentUpdation" :condition="condition">
      <input type="text" v-model="condition.argument" />
    </slot>
    <slot
      name="conditionDeletion"
      :deleteCondition="() => $emit('deleteCondition', condition)"
    >
      <button @click="$emit('deleteCondition', condition)">x</button>
    </slot>
  </div>
</template>
