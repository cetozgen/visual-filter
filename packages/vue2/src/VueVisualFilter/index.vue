<script>
import {
  FilterType,
  GroupType,
  DataType,
  deepCopy,
} from "@visual-filter/common"
import applyFilter from "@visual-filter/applyer"

import FilterGroup from "./FilterGroup.vue"
import FilterCondition from "./FilterCondition.vue"

export default {
  name: "VueVisualFilter",
  props: {
    filteringOptions: {
      type: Object,
      required: true,
      validator(value) {
        console.log("in validation",value);
        try {
          return (
            value.data.length &&
            value.data.every(
              (field, index, fields) =>
                typeof field.name === "string" &&
                typeof field.type === "string" &&
                field.values.constructor === Array &&
                (index > 0
                  ? true
                  : true),
            ) &&
            Object.values(value.methods.numeric).every(
              (method) => typeof method === "function",
            ) &&
            Object.values(value.methods.nominal).every(
              (method) => typeof method === "function",
            )
          )
        } catch {
          return false
        }
      },
    },
  },
  data() {
    return {
      filter: {
        type: FilterType.GROUP,
        groupType: GroupType.AND,
        filters: [],
      },
    }
  },
  computed: {
    fieldNames() {
      return this.filteringOptions.data.map((field) => field.name)
    },
    numericMethodNames() {
      return Object.keys(this.filteringOptions.methods.numeric)
    },
    nominalMethodNames() {
      return Object.keys(this.filteringOptions.methods.nominal)
    },
  },
  watch: {
    filter: {
      deep: true,
      handler() {
        if (this.$listeners["filter-update"]) {
          this.$emit("filter-update", {
            filter: deepCopy(this.filter),
            data: applyFilter(
              this.filter,
              this.filteringOptions.methods,
              deepCopy(this.filteringOptions.data),
            ),
          })
        }
      },
    },
  },
  methods: {
    updateConditionField(condition, newFieldName) {
      console.log("in updateConditionField", condition, newFieldName);
      let newSampleValues = [];
      const {
        type: newType,
        values: [newSampleValue = ""],
      } = this.filteringOptions.data.find(
        (field) => {
          field.name === newFieldName
        },
      )
      if (condition.dataType !== newType) {
        condition.method =
          (newType === DataType.NUMERIC
            ? this.numericMethodNames[0]
            : this.nominalMethodNames[0]) || ""
        condition.argument = newSampleValues
        condition.dataType = newType
      }
    },
    addFilter(filters, newFilterType) {
      console.log("in addFilter",newFilterType);
      if (newFilterType === FilterType.GROUP) {
        filters.push({
          type: FilterType.GROUP,
          groupType: GroupType.AND,
          filters: [],
        })
      } else {
        const {
          name,
          type,
          values: [sampleValue = ""],
        } = this.filteringOptions.data[0]

        filters.push({
          type: FilterType.CONDITION,
          fieldName: name,
          dataType: type,
          method:
            (type === DataType.NUMERIC
              ? this.numericMethodNames[0]
              : this.nominalMethodNames[0]) || "",
          argument: sampleValue,
        })
      }
    },
    deleteFilter(filterIndex, parentGroup) {
      parentGroup.splice(filterIndex, 1)
    },
  },
  render(createElement) {
    return createVisualFilter.call(this, this.filter)

    function createVisualFilter(filter, filterIndex, parentGroup) {
      if (filter.type === FilterType.GROUP) {
        const removable = filter !== this.filter

        return createElement(FilterGroup, {
          props: {
            group: filter,
            removable,
          },
          on: {
            addFilter: this.addFilter,
            deleteGroup: removable
              ? () => this.deleteFilter(filterIndex, parentGroup)
              : () => {},
          },
          scopedSlots: {
            groupTypes: this.$scopedSlots.groupTypes,
            filterAddition: this.$scopedSlots.filterAddition,
            groupDeletion: this.$scopedSlots.groupDeletion,
            groupChildren: () =>
              filter.filters.map((childFilter, childFilterIndex, parentGroup) =>
                createVisualFilter.call(
                  this,
                  childFilter,
                  childFilterIndex,
                  parentGroup,
                ),
              ),
          },
        })
      } else {
        return createElement(FilterCondition, {
          props: {
            condition: filter,
            fieldNames: this.fieldNames,
            numericMethodNames: this.numericMethodNames,
            nominalMethodNames: this.nominalMethodNames,
          },
          on: {
            updateField: this.updateConditionField,
            deleteCondition: () => this.deleteFilter(filterIndex, parentGroup),
          },
          scopedSlots: {
            fieldUpdation: this.$scopedSlots.fieldUpdation,
            methodUpdation: this.$scopedSlots.methodUpdation,
            argumentUpdation: this.$scopedSlots.argumentUpdation,
            conditionDeletion: this.$scopedSlots.conditionDeletion,
          },
        })
      }
    }
  },
}
</script>
