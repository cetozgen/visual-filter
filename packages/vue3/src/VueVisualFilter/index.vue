<script>
import { h } from "vue"
import {
  FilterType,
  GroupType,
  DataType,
  deepCopy,
} from "@visual-filter/common"
import applyFilter from "@visual-filter/applyer"

import FilterGroup from "./FilterGroup.vue"
import FilterCondition from "./FilterCondition.vue"

export default {
  name: "VueVisualFilter",
  emits: ["filterUpdate"],
  props: {
    filteringOptions: {
      type: Object,
      required: true,
      validator(value) {
        try {
          return (
            value.data.length &&
            value.data.every(
              (field, index, fields) =>
                typeof field.name === "string" &&
                typeof field.type === "string" &&
                field.values.constructor === Array &&
                (index > 0
                  ? field.values.length === fields[index - 1].values.length
                  : true),
            ) &&
            Object.values(value.methods.numeric).every(
              (method) => typeof method === "function",
            ) &&
            Object.values(value.methods.nominal).every(
              (method) => typeof method === "function",
            )
          )
        } catch {
          return false
        }
      },
    },
  },
  data() {
    return {
      filter: {
        type: FilterType.GROUP,
        groupType: GroupType.AND,
        filters: [],
      },
    }
  },
  computed: {
    fieldNames() {
      return this.filteringOptions.data.map((field) => field.name)
    },
    numericMethodNames() {
      return Object.keys(this.filteringOptions.methods.numeric)
    },
    nominalMethodNames() {
      return Object.keys(this.filteringOptions.methods.nominal)
    },
  },
  watch: {
    filter: {
      deep: true,
      handler() {
        this.$emit("filterUpdate", {
          filter: deepCopy(this.filter),
          data: applyFilter(
            this.filter,
            this.filteringOptions.methods,
            deepCopy(this.filteringOptions.data),
          ),
        })
      },
    },
  },
  methods: {
    updateConditionField(condition, newFieldName) {
      const {
        type: newType,
        values: [newSampleValue = ""],
      } = this.filteringOptions.data.find(
        (field) => field.name === newFieldName,
      )
      if (condition.dataType !== newType) {
        condition.method =
          (newType === DataType.NUMERIC
            ? this.numericMethodNames[0]
            : this.nominalMethodNames[0]) || ""
        condition.argument = newSampleValue
        condition.dataType = newType
      }
    },
    addFilter(filters, newFilterType) {
      if (newFilterType === FilterType.GROUP) {
        filters.push({
          type: FilterType.GROUP,
          groupType: GroupType.AND,
          filters: [],
        })
      } else {
        const {
          name,
          type,
          values: [sampleValue = ""],
        } = this.filteringOptions.data[0]

        filters.push({
          type: FilterType.CONDITION,
          fieldName: name,
          dataType: type,
          method:
            (type === DataType.NUMERIC
              ? this.numericMethodNames[0]
              : this.nominalMethodNames[0]) || "",
          argument: sampleValue,
        })
      }
    },
    deleteFilter(filterToDelete) {
      function recursiveDeletion(filter, index, filters) {
        if (filter === filterToDelete) {
          filters.splice(index, 1)
        } else if (filter.type === FilterType.GROUP) {
          filter.filters.map(recursiveDeletion)
        }
      }

      if (filterToDelete !== this.filter) {
        recursiveDeletion(this.filter)
      }
    },
  },
  render() {
    const createVisualizer = (filter) => {
      if (filter.type === FilterType.GROUP) {
        return h(
          FilterGroup,
          {
            group: filter,
            filterTypes: Object.values(FilterType),
            groupTypes: Object.values(GroupType),
            removable: filter !== this.filter,
            onAddFilter: this.addFilter,
            onDeleteGroup: this.deleteFilter,
          },
          {
            groupTypes: this.$slots.groupTypes,
            filterAddition: this.$slots.filterAddition,
            groupDeletion: this.$slots.groupDeletion,
            groupChildren: () => filter.filters.map(createVisualizer),
          },
        )
      } else {
        return h(
          FilterCondition,
          {
            condition: filter,
            fieldNames: this.fieldNames,
            numericMethodNames: this.numericMethodNames,
            nominalMethodNames: this.nominalMethodNames,
            onUpdateField: this.updateConditionField,
            onDeleteCondition: this.deleteFilter,
          },
          {
            fieldUpdation: this.$slots.fieldUpdation,
            methodUpdation: this.$slots.methodUpdation,
            argumentUpdation: this.$slots.argumentUpdation,
            conditionDeletion: this.$slots.conditionDeletion,
          },
        )
      }
    }

    return createVisualizer(this.filter)
  },
}
</script>
