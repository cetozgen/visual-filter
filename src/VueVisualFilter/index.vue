<script>
import { h } from "vue"
import FilterGroup from "./FilterGroup.vue"
import FilterCondition from "./FilterCondition.vue"
import { createEnum, deepCopy, applyFilter } from "@/helpers.js"

export const FilterType = createEnum({
  GROUP: "group",
  CONDITION: "condition"
})

export const GroupType = createEnum({
  AND: "and",
  NOT_AND: "not and",
  OR: "or",
  NOT_OR: "not or"
})

export const DataType = createEnum({
  NUMERIC: "numeric",
  NOMINAL: "nominal"
})

export default {
  name: "VisualFilter",
  emits: ["filterUpdate"],
  props: {
    filteringOptions: {
      type: Object,
      required: true,
      validator(value) {
        try {
          return (
            value.data.length &&
            value.data.every(
              (field) =>
                typeof field.name === "string" &&
                typeof field.type === "string" &&
                field.values.constructor === Array
            ) &&
            value.methods.numeric &&
            Object.values(value.methods.numeric).every(
              (method) => typeof method === "function"
            ) &&
            Object.values(value.methods.nominal).every(
              (method) => typeof method === "function"
            )
          )
        } catch {
          return false
        }
      }
    }
  },
  data() {
    return {
      filter: {
        type: FilterType.GROUP,
        groupType: GroupType.AND,
        filters: []
      }
    }
  },
  computed: {
    fieldNames() {
      return this.filteringOptions.data.map((field) => field.name)
    },
    numericMethodNames() {
      return Object.keys(this.filteringOptions.methods.numeric)
    },
    nominalMethodNames() {
      return Object.keys(this.filteringOptions.methods.nominal)
    }
  },
  watch: {
    filter: {
      deep: true,
      handler() {
        this.$emit("filterUpdate", {
          filter: deepCopy(this.filter),
          data: applyFilter(
            this.filter,
            this.filteringOptions.methods,
            deepCopy(this.filteringOptions.data)
          )
        })
      }
    }
  },
  methods: {
    updateConditionField(condition, newFieldName) {
      const {
        type,
        values: [sampleValue]
      } = this.filteringOptions.data.find(
        (field) => field.name === newFieldName
      )
      condition.fieldName = newFieldName
      condition.dataType = type
      condition.method =
        type === DataType.NUMERIC
          ? this.numericMethodNames[0]
          : this.nominalMethodNames[0]
      condition.argument = sampleValue
    },
    updateConditionMethod(condition, newMethodName) {
      condition.method = newMethodName
    },
    updateConditionArgument(condition, newArgumentValue) {
      condition.argument = newArgumentValue
    },
    updateGroupType(group, newGroupType) {
      group.groupType = newGroupType
    },
    addFilter(filters, newFilterType) {
      if (newFilterType === FilterType.GROUP) {
        filters.push({
          type: FilterType.GROUP,
          groupType: GroupType.AND,
          filters: []
        })
      } else {
        const {
          name,
          type,
          values: [sampleValue]
        } = this.filteringOptions.data[0]

        filters.push({
          type: FilterType.CONDITION,
          fieldName: name,
          dataType: type,
          method:
            type === DataType.NUMERIC
              ? this.numericMethodNames[0]
              : this.nominalMethodNames[0],
          argument: sampleValue
        })
      }
    },
    deleteFilter(filterToDelete) {
      function recursiveDeletion(filter, index, filters) {
        if (filter === filterToDelete) {
          filters.splice(index, 1)
        } else if (filter.type === FilterType.GROUP) {
          filter.filters.map(recursiveDeletion)
        }
      }

      if (filterToDelete !== this.filter) {
        recursiveDeletion(this.filter)
      }
    }
  },
  render() {
    const createVisualizer = (filter) => {
      if (filter.type === FilterType.GROUP) {
        return h(
          FilterGroup,
          {
            group: filter,
            filterTypes: Object.values(FilterType),
            groupTypes: Object.values(GroupType),
            removable: filter !== this.filter,
            onUpdateType: this.updateGroupType,
            onAddFilter: this.addFilter,
            onDeleteGroup: this.deleteFilter
          },
          () => filter.filters.map(createVisualizer)
        )
      } else {
        return h(FilterCondition, {
          condition: filter,
          fieldNames: this.fieldNames,
          numericMethodNames: this.numericMethodNames,
          nominalMethodNames: this.nominalMethodNames,
          onUpdateField: this.updateConditionField,
          onUpdateMethod: this.updateConditionMethod,
          onUpdateArgument: this.updateConditionArgument,
          onDeleteCondition: this.deleteFilter
        })
      }
    }

    return createVisualizer(this.filter)
  }
}
</script>
