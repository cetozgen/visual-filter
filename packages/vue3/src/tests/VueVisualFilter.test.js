import { render, fireEvent } from "@testing-library/vue"
import "@testing-library/jest-dom"
import { GroupType, FilterType, DataType } from "@visual-filter/common"

import VueVisualFilter from "@/VueVisualFilter/index.vue"

beforeEach((ctx) => {
  ctx.commonProps = {
    filteringOptions: {
      data: [
        {
          name: "First Name",
          type: "nominal",
          values: ["Foo", "Fizz"],
        },
        {
          name: "Last Name",
          type: "nominal",
          values: ["Bar", "Buzz"],
        },
        {
          name: "Age",
          type: "numeric",
          values: [18, 19],
        },
      ],
      methods: {
        numeric: {
          "="(cellValue, argument) {
            return cellValue == argument
          },
          ">"(cellValue, argument) {
            return cellValue > argument
          },
        },
        nominal: {
          contains(cellValue, argument) {
            return cellValue.includes(argument)
          },
        },
      },
    },
  }
})

test("`filteringOptions` prop validator should work as expected", ({
  commonProps,
}) => {
  const filteringOptionsValidator =
    VueVisualFilter.props.filteringOptions.validator

  expect(filteringOptionsValidator(undefined)).toBeFalsy()

  expect(filteringOptionsValidator(null)).toBeFalsy()

  expect(filteringOptionsValidator({})).toBeFalsy()

  expect(filteringOptionsValidator({ data: [] })).toBeFalsy()

  expect(filteringOptionsValidator({ data: [{}] })).toBeFalsy()

  expect(
    filteringOptionsValidator({ data: [{ name: "Firstname" }] }),
  ).toBeFalsy()

  expect(
    filteringOptionsValidator({
      data: [{ name: "Firstname", type: "nominal" }],
    }),
  ).toBeFalsy()

  expect(
    filteringOptionsValidator({
      data: [{ name: "Firstname", type: "nominal", values: [] }],
    }),
  ).toBeFalsy()

  expect(
    filteringOptionsValidator({
      data: [{ name: "Firstname", type: "nominal", values: ["Foo"] }],
    }),
  ).toBeFalsy()

  expect(
    filteringOptionsValidator({
      data: [
        { name: "Firstname", type: "nominal", values: ["a", "b"] },
        { name: "Firstname", type: "nominal", values: ["x"] },
      ],
    }),
  ).toBeFalsy()

  expect(
    filteringOptionsValidator({
      data: [{ name: "Firstname", type: "nominal", values: ["a"] }],
      methods: {
        nominal: null,
      },
    }),
  ).toBeFalsy()

  expect(
    filteringOptionsValidator({
      data: [{ name: "Firstname", type: "nominal", values: ["a"] }],
      methods: {
        nominal: { a: null },
      },
    }),
  ).toBeFalsy()

  expect(
    filteringOptionsValidator({
      data: [{ name: "Firstname", type: "nominal", values: ["a"] }],
      methods: {
        nominal: {
          contian: () => {},
        },
      },
    }),
  ).toBeFalsy()

  // TODO: should be handled
  // expect(
  //   filteringOptionsValidator({
  //     data: [{ name: "Firstname", type: "nominal", values: ["a"] }],
  //     methods: {
  //       nominal: {
  //         contian: () => {},
  //       },
  //       numeric: 0,
  //     },
  //   }),
  // ).toBeFalsy()

  expect(
    filteringOptionsValidator({
      data: [{ name: "Firstname", type: "nominal", values: ["a"] }],
      methods: {
        nominal: {
          contian: () => {},
        },
        numeric: {
          a: null,
        },
      },
    }),
  ).toBeFalsy()

  expect(filteringOptionsValidator(commonProps.filteringOptions)).toBeTruthy()
})

test("after initial render we should have group type and filter type selects with correct initial options", ({
  commonProps,
}) => {
  const methods = render(VueVisualFilter, {
    props: commonProps,
  })

  const groupTypeSelect = methods.queryByTestId("group-type-select")

  expect(groupTypeSelect).toBeTruthy()
  expect(groupTypeSelect).toHaveValue(GroupType.AND)

  const filterTypeSelect = methods.queryByTestId("filter-type-select")

  expect(filterTypeSelect).toBeTruthy()
  expect(filterTypeSelect).toHaveValue(FilterType.GROUP)
})

test("expect `onFilterUpdate` event to be called with correct parameters when changing group type select", async ({
  commonProps,
}) => {
  const spyFilterUpdateEvent = vi.fn()
  const methods = render(VueVisualFilter, {
    propsData: { ...commonProps, onFilterUpdate: spyFilterUpdateEvent },
  })

  const groupTypeSelect = methods.queryByTestId("group-type-select")

  await fireEvent.update(groupTypeSelect, GroupType.NOT_AND)
  expect(spyFilterUpdateEvent.calls).toMatchSnapshot()

  await fireEvent.update(groupTypeSelect, GroupType.OR)
  expect(spyFilterUpdateEvent.calls).toMatchSnapshot()

  await fireEvent.update(groupTypeSelect, GroupType.NOT_OR)
  expect(spyFilterUpdateEvent.calls).toMatchSnapshot()
})

test("filter of type condition renders as expected on initial render after selecting from group type select, and callback is called with correct parameters", async ({
  commonProps,
}) => {
  const spyFilterUpdateEvent = vi.fn()
  const methods = render(VueVisualFilter, {
    props: {
      ...commonProps,
      onFilterUpdate: spyFilterUpdateEvent,
    },
  })

  const filterTypeSelect = methods.queryByTestId("filter-type-select")

  await fireEvent.update(filterTypeSelect, FilterType.CONDITION)

  const {
    data: [
      {
        fieldName: firstFieldName,
        type: firstFieldType,
        values: [firstValue],
      },
    ],
    methods: methodsOptions,
  } = commonProps.filteringOptions

  const fieldNameSelect = methods.queryByTestId("field-name-select")

  expect(fieldNameSelect).toBeTruthy()
  expect(fieldNameSelect).toHaveValue(firstFieldName)

  const methodSelect = methods.queryByTestId("method-select")

  expect(methodSelect).toBeTruthy()
  expect(methodSelect).toHaveValue(
    Object.keys(
      firstFieldType === DataType.NOMINAL
        ? methodsOptions.nominal
        : methodsOptions.numeric,
    )[0],
  )

  const argumentInput = methods.queryByTestId("argument-input")

  expect(argumentInput).toBeTruthy()
  expect(argumentInput).toHaveValue(firstValue)

  expect(spyFilterUpdateEvent.calls).toMatchSnapshot()
})
