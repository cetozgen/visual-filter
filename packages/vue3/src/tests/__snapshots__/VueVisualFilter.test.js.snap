// Vitest Snapshot v1

exports[`expect \`onFilterUpdate\` event to be called with correct parameters when changing group type select 1`] = `
[
  [
    {
      "data": [
        {
          "name": "First Name",
          "type": "nominal",
          "values": [
            "Foo",
            "Fizz",
          ],
        },
        {
          "name": "Last Name",
          "type": "nominal",
          "values": [
            "Bar",
            "Buzz",
          ],
        },
        {
          "name": "Age",
          "type": "numeric",
          "values": [
            18,
            19,
          ],
        },
      ],
      "filter": {
        "filters": [],
        "groupType": "not and",
        "type": "group",
      },
    },
  ],
]
`;

exports[`expect \`onFilterUpdate\` event to be called with correct parameters when changing group type select 2`] = `
[
  [
    {
      "data": [
        {
          "name": "First Name",
          "type": "nominal",
          "values": [
            "Foo",
            "Fizz",
          ],
        },
        {
          "name": "Last Name",
          "type": "nominal",
          "values": [
            "Bar",
            "Buzz",
          ],
        },
        {
          "name": "Age",
          "type": "numeric",
          "values": [
            18,
            19,
          ],
        },
      ],
      "filter": {
        "filters": [],
        "groupType": "not and",
        "type": "group",
      },
    },
  ],
  [
    {
      "data": [
        {
          "name": "First Name",
          "type": "nominal",
          "values": [],
        },
        {
          "name": "Last Name",
          "type": "nominal",
          "values": [],
        },
        {
          "name": "Age",
          "type": "numeric",
          "values": [],
        },
      ],
      "filter": {
        "filters": [],
        "groupType": "or",
        "type": "group",
      },
    },
  ],
]
`;

exports[`expect \`onFilterUpdate\` event to be called with correct parameters when changing group type select 3`] = `
[
  [
    {
      "data": [
        {
          "name": "First Name",
          "type": "nominal",
          "values": [
            "Foo",
            "Fizz",
          ],
        },
        {
          "name": "Last Name",
          "type": "nominal",
          "values": [
            "Bar",
            "Buzz",
          ],
        },
        {
          "name": "Age",
          "type": "numeric",
          "values": [
            18,
            19,
          ],
        },
      ],
      "filter": {
        "filters": [],
        "groupType": "not and",
        "type": "group",
      },
    },
  ],
  [
    {
      "data": [
        {
          "name": "First Name",
          "type": "nominal",
          "values": [],
        },
        {
          "name": "Last Name",
          "type": "nominal",
          "values": [],
        },
        {
          "name": "Age",
          "type": "numeric",
          "values": [],
        },
      ],
      "filter": {
        "filters": [],
        "groupType": "or",
        "type": "group",
      },
    },
  ],
  [
    {
      "data": [
        {
          "name": "First Name",
          "type": "nominal",
          "values": [],
        },
        {
          "name": "Last Name",
          "type": "nominal",
          "values": [],
        },
        {
          "name": "Age",
          "type": "numeric",
          "values": [],
        },
      ],
      "filter": {
        "filters": [],
        "groupType": "not or",
        "type": "group",
      },
    },
  ],
]
`;

exports[`filter of type condition renders as expected on initial render after selecting from group type select, and callback is called with correct parameters 1`] = `
[
  [
    {
      "data": [
        {
          "name": "First Name",
          "type": "nominal",
          "values": [
            "Foo",
          ],
        },
        {
          "name": "Last Name",
          "type": "nominal",
          "values": [
            "Bar",
          ],
        },
        {
          "name": "Age",
          "type": "numeric",
          "values": [
            18,
          ],
        },
      ],
      "filter": {
        "filters": [
          {
            "argument": "Foo",
            "dataType": "nominal",
            "fieldName": "First Name",
            "method": "contains",
            "type": "condition",
          },
        ],
        "groupType": "and",
        "type": "group",
      },
    },
  ],
]
`;
